062-topic-replication-factor-change.md

```diff
@@ -0,0 +1,254 @@
+This is is far from being user friendly, and the user is in charge of doing the calculations to determine the best broker to host a new replica, or which replica needs to be dropped.
+
+When data loss is not a problem, the user can simply delete and recreate the `KafkaTopic` resource with the desired replication factor.
+This is not really an option for most use cases.
+
+## Motivation
+
+The topic replication factor change feature will provide a much better cloud-native experience for Strimzi users.
+They will be able to change the topic replication factor by simply updating the `KafkaTopic` resource, which may also happen through a GitOps pipeline.
+
+This is a non exhaustive list of use cases that will become significantly easier:
+
+- Set a lower replication factor for non-critical topics to reduce the infrastructure costs (compute, storage and network).
+- Set a lower replication factor to deal with resource shortage (trade-off between fault-tolerance and resource utilization).
+- Set a higher replication factor for critical topics to improve durability and fault tolerance.
+
+## Proposal
+
+The replication factor change feature can be implemented by integrating the Topic Operator with Cruise Control.
+The Cluster Operator is already integrated with Cruise Control, so we can reuse some of the existing logic and classes.
+
+At runtime, the Topic Operator watches and periodically reconcile the configuration of all managed and non-paused `KafkaTopic` resources.
+Among the other configurations, the reconciliation logic also search for `.spec.replicas` update by comparing with the actual topic replication factor.
+From the point of view of the Topic Operator, a topic replication factor change can be in the following states:
+
+- **Pending**: `.spec.replicas` is updated, but the replication factor change is not sent to Cruise Control.
+- **Ongoing**: `.spec.replicas` is updated and Cruise Control has accepted the request, but the change is not completed, or there is some error.
+- **Ready**: `.spec.replicas` is updated and Cruise Control execution is completed (no warning in `KafkaTopic` status).
+
+Only one replication factor change can be ongoing at any given time for a specific `KafkaTopic`, additional changes for the same `KafkaTopic` will remain pending, and periodically retried.
+At most `max.active.user.tasks` replication factor changes for different `KafkaTopic` resources can be ongoing at the same time in Cruise Control (5 by default).
+If required, this limit can be increased from the `Kafka` resource using the `.spec.cruiseControl.config` field.
+When the task queue is full, following replication factor changes will fail with a warning, and periodically retried until the queue frees up.
+
+During the whole process, the `KafkaTopic` remains in a `Ready` state, so we use an additional `Warning` condition to communicate the replication factor change status.
```

Comment:
I am not sure "Warning" is kind of right for an ongoing operation. Why not a dedicated field in the `.status` section? In general, a "warning" condition let me think about ... a warning ... something which should be fixed, not a real blocker error but something that I should be worried about, not an ongoing operation about what I asked for.

Reply:
I tend to agree, but that's tricky with CC. I tried to clarify what I mean in a previous answer to one of your comments. Let me know what you think. Thanks.